samples/matter/common/src/event_triggers/default_event_triggers.cpp

```diff
@@ -20,19 +21,68 @@ using namespace chip;
 
 namespace
 {
-k_timer sFactoryResetDelayTimer;
-
-void FactoryResetDelayTimerTimeoutCallback(k_timer *timer)
+/* Timer to delay some operations to allow send back ACK to Matter Controller. */
+k_timer sDelayTimer;
+
+/* Some operations may interrupt communication with the Matter controller so we need to postpone the action because it
+ * must receive the acknowledgement from the device.  */
+enum class DelayedAction { FactoryReset, BlockMainThread };
+struct DelayedCtx {
+	DelayedAction action;
+	TestEventTrigger::TriggerValue value;
+};
+
+void DelayTimerCallback(k_timer *timer)
 {
-	Server::GetInstance().ScheduleFactoryReset();
+	/* Schedule an action in the Main thread to allow releasing Delayed context memory (it cannot be done within the
+	 * Timer IRQ)*/
+	Nrf::PostTask([timer] {
+		Platform::UniquePtr<DelayedCtx> ctx(reinterpret_cast<DelayedCtx *>(k_timer_user_data_get(timer)));
+		if (!ctx) {
+			return;
+		}
+
+		TestEventTrigger::TriggerValue value = ctx->value;
+
+		switch (ctx->action) {
+		case DelayedAction::FactoryReset:
+			Server::GetInstance().ScheduleFactoryReset();
+			break;
+		case DelayedAction::BlockMainThread:
+			Nrf::PostTask([value] {
+				int64_t startTime = k_uptime_get();
+
+				LOG_INF("Blocking Main thread for %d ms", value);
+
+				while (startTime + value > k_uptime_get()) {
+					;
+				}
+
+				LOG_INF("Blocking Main thread finished");
+			});
+			break;
+		default:
+			break;
+		}
+	});
 }
 
 CHIP_ERROR FactoryResetCallback(TestEventTrigger::TriggerValue delayMs)
 {
 	LOG_DBG("Called Factory Reset event trigger with delay %d ms", delayMs);
 
-	k_timer_init(&sFactoryResetDelayTimer, &FactoryResetDelayTimerTimeoutCallback, nullptr);
-	k_timer_start(&sFactoryResetDelayTimer, K_MSEC(delayMs), K_NO_WAIT);
+	Platform::UniquePtr<DelayedCtx> delayedContext(Platform::New<DelayedCtx>());
+	if (!delayedContext) {
+		return CHIP_ERROR_NO_MEMORY;
+	}
+
+	delayedContext->action = DelayedAction::FactoryReset;
+	delayedContext->value = delayMs;
+
+	k_timer_start(&sDelayTimer, K_MSEC(delayMs), K_NO_WAIT);
+	k_timer_user_data_set(&sDelayTimer, reinterpret_cast<void *>(delayedContext.get()));
```

Comment:
If two delayed actions are executed close to each other, this will result in a leak. If we really need heap allocated context for the timer, we must make sure it's expired before spawning a new timer.

Reply:
Ok, I will think about protection or eliminating heap allocation.

Reply:
I've added a protection there.